Centroid atomic stress for shake, rattle and rigid/small

donatas-surblys opened the pull request:
**Summary**

This implements the centroid form atomic stress for constrained dynamics via shake and rattle fixes, as well as rigid body dynamics for fix rigid/small. The centroid form provides strictly correct heat flux values when used with the heat/flux compute for the whole system, and gives good approximation of local heat flux over control volumes. Without the centroid form, the original implementation gives heat flux values that don't quite match the input/output of thermal baths, like langevin and ehex fixes. Details can be found in our recently published paper [1].

If the changes look OK, I can also go on to update the documentation.

**Related Issue(s)**

This is a continuation of work done in  PR #1704.

**Author(s)**

Donatas Surblys (Tohoku University) surblys.donatas at gmail dot com

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

This might change the heat flux values obtained from shake, rattle, and rigid/small fixes if somebody was trying to use centroid/stress/atom and LAMMPS didn't properly error out.

**Implementation Notes**

New v_tally functions are defined for shake/rigid and rigid/small in fix.cpp, that internally call the original v_tally functions, and then do additional centroid stress computations if necessary.

For shake/rigid:
Constraint forces are treated as pair interactions and handled respectively.

For rigid/small:
The Body structure in fix_rigid_small.h is updated to the following variables:
```
    int natoms;            // total number of atoms in body
    double xgc[3];         // geometric center position
    double xgc_body[3];    // geometric center relative to xcm in body coords
```
This allows computation of the centroid position, enabling easily obtaining centroid form atomic stress, even when not all atoms are owned by the process. I'd like some feedback if I approached in an acceptable way, and have some slight doubts if wrapping was handled correctly.

What's still missing:
- fix rigid: Haven't really looked into it, but might be possible to use same approach as fix rigid/small. Don't know if there's much need.
- TIP4P: Need quite an involved approach. A naive implementation was not successful.
- Many-body potentials.

Notes about the original implementation of rigid/small:
Currently, fix rigid/small gives unphysical results for y an z components when used with the heat/flux compute. This is because the atomic stress virial is not averaged over all atoms in the molecule, i.e. the atomic stress tensor is not symmetric, but only the upper triangular is stored. Because of this

> jv[0]  = rijx fijx vjx + rijx fijy vjy + rijx fijz vjz
> jv[1]  = rijy fijx vjx + rijy fijy vjy + rijy fijz vjz
> jv[2]  = rijz fijx vjx + rijz fijy vjy + rijz fijz vjz

becomes

> jv[0]  = rijx fijx vjx + rijx fijy vjy + rijx fijz vjz
> jv[1]  = rijx fijy vjx + rijy fijy vjy + rijy fijz vjz
> jv[2]  = rijx fijz vjx + rijy fijz vjy + rijz fijz vjz

As you can see, only the x component is correct. In practice, the other components give obviously unphysical values.
I strongly believe there should be at least a warning cautioning against such use.

**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

[1] https://doi.org/10.1063/5.0070930

athomps reviewed (approved):
I approve. Nice work!

akohlmey commented on src/RIGID/fix_rigid_small.cpp:
```diff
@@ -73,6 +73,7 @@ FixRigidSmall::FixRigidSmall(LAMMPS *lmp, int narg, char **arg) :
   dof_flag = 1;
   enforce2d_flag = 1;
   stores_ids = 1;
+  centroidstressflag = CENTROID_AVAIL;
```
This flag will be inherited by the `FixRigidNHSmall` class and thus the corresponding derived classes, but I don't see the corresponding changes to enable the feature; similarly `FixRigidSmallOMP`. So it either needs to be disabled in the corresponding constructors or the equivalent changes added.

akohlmey reviewed (changes requested):
As mentioned in the PR description, updated docs are missing and the suggested warning should be added.

donatas-surblys commented:
Regarding, `FixRigidNHSmall`, I think it's OK as is,
because the modifications in `FixRigidSmall` were in the `setup`, `set_xv`, `set_v`
and various communication functions, which are all called either explicitly or
through inheritance by `FixRigidNHSmall`. Actually, our paper was done with the
`fix rigid/nve/small`, which I believe inherits from `FixRigidNHSmall`.
Regarding the OMP version, yeah, I missed it, as it didn't come out with "-sf omp"
when using `fix rigid/nve/small`.
I'd like to fix it up too eventually, but I'll disable it for the time being.
I've also disabled centroid stress for non-small rigid fixes.

I was looking into having LAMMPS emit a warning about using `compute heat/flux`
with standard atomic virial from `fix rigid` and `fix rigid/small`.
Unfortunately, I can't seem to think up a good way, as the warning message should be
emitted from `compute_heat_flux.cpp`, but at that point there's not enough
information about the underlying fixes that were used to compute the virial.
I guess it would be easiest to add a new flag to the `stress/atom` compute,
but this doesn't seem like a clean approach.
For the time being, I've added a separate warning to the manual.

I've added short explanations and a new citation to the manual.
I've also added a short clarification on how the standard virial is computed
for `fix rigid`.

I've also added citation information to the centroid/stress/atom,
as has been suggested in the past.
